File: 0000-0999/0046.Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        # return list(itertools.permutations(nums))
        ans = []
        self.dfs(nums, [], ans)
        return ans

    def dfs(self, nums, temp, ans):
        if not nums:
            ans.append(temp)
            return
        for i in xrange(len(nums)):
            self.dfs(nums[:i]+nums[i+1:], temp+[nums[i]], ans)
            '''
            nums[0], nums[i] = nums[i], nums[0]
            self.dfs(nums[1:], temp+[nums[0]], ans)
            '''

File: 0000-0999/0046.Permutations.go
func permute(nums []int) [][]int {
    n := len(nums)
    perm := make([]int, n)
    used := make([]bool, n)
    ans := [][]int{}
    dfs(nums, perm, used, n, 0, &ans)
    return ans
}

func dfs(nums, perm []int, used []bool, n, cur int, ans *[][]int) {
    if cur == n {
        tmp := make([]int, n)
        copy(tmp, perm)
        *ans = append(*ans, tmp)
        return
    }
    for i := 0; i < n; i++ {
        if used[i] {
            continue
        }
        perm[cur] = nums[i]
        used[i] = true
        dfs(nums, perm, used, n, cur+1, ans)
        used[i] = false
    }
}